include/portability.h

```diff
@@ -2,6 +2,15 @@
 #define CBITSET_PORTABILITY_H
 #include <stdint.h>
 
+// For compatibility with MSVC with the use of `restrict`
+#if defined(__INTEL_COMPILER) || defined(__GNUC__) || defined(__GNUG__) || defined(__clang__) || (__STDC_VERSION__ >= 199901L)
+#define CBITSET_RESTRICT restrict
+#elif defined(_MSC_VER)
+#define CBITSET_RESTRICT __restrict
+#else
+#define RESTRICT
+#endif // defined(__INTEL_COMPILER) || defined(__GNUC__) || defined(__GNUG__) || defined(__clang__) || (__STDC_VERSION__ >= 199901L)
```

Comment:
Sorry for the drive-by pedantry from someone with no relationship to the project; feel free to ignore me!

I see a few problems with this if you care about portability.

 * This will result in errors from C++. `restrict` isn't supported, though `__restrict` is in most of them (in both C and C++ modes).  `restrict` should really be restricted to C99+ compilers w/o regard to the vendor
 * icl (ICC's MSVC-compatible frontend) won't work here as it doesn't support `restrict` but does define `__INTEL_COMPILER`.
 * `defined(__GNUG__)` doesn't do any good here; `__GNUC__` is defined in both C and C++ mode, `defined(__GNUG__)` is really just a shortcut for `defined(__GNUC__) && defined(__cplusplus)`

There is a [`HEDLEY_RESTRICT`](https://github.com/nemequ/hedley/blob/master/hedley.h#L1440-L1465) macro in Hedley which has some logic for this which you're welcome to steal (it's public domain). Hedley is one of my projects, though, so take the suggestion with a grain of salt :)

Reply:
Thanks Evan. I know of your work. This is a C library. Is hedley C compliant? If so, we could just add it to this project and be done with the whole issue.

Reply:
Hedley is usable from C, yes.  Honestly, I'd say it's C first but also works in C++.

Technically I guess it's not really C "compliant"; it will take advantage of compiler-specific features (such as `__restrict`) when available, so it doesn't really preprocess to standards-compliant C when it knows the compiler can do better. But I don't think that's what you meant.